Installation issue: mysql

### Steps to reproduce the issue

```console
$ spack spec mysql
...
-   mysql@8.0.29%gcc@11.2.0~client_only~ipo build_system=cmake build_type=Release cxxstd=14 generator=make arch=linux-centos7-haswell
[+]      ^bison@3.8.2%gcc@11.2.0 build_system=autotools arch=linux-centos7-haswell
[+]          ^diffutils@3.9%gcc@11.2.0 build_system=autotools arch=linux-centos7-haswell
[+]          ^gettext@0.22.3%gcc@11.2.0+bzip2+curses+git~libunistring+libxml2+pic+shared+tar+xz build_system=autotools arch=linux-centos7-haswell
[+]              ^libxml2@2.10.3%gcc@11.2.0+pic~python+shared build_system=autotools arch=linux-centos7-haswell
[+]              ^tar@1.34%gcc@11.2.0 build_system=autotools zip=pigz arch=linux-centos7-haswell
[+]                  ^pigz@2.7%gcc@11.2.0 build_system=makefile arch=linux-centos7-haswell
[+]          ^m4@1.4.19%gcc@11.2.0+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=linux-centos7-haswell
[+]              ^libsigsegv@2.14%gcc@11.2.0 build_system=autotools arch=linux-centos7-haswell
[+]          ^perl@5.38.0%gcc@11.2.0+cpanm+opcode+open+shared+threads build_system=generic patches=714e4d1 arch=linux-centos7-haswell
[+]              ^berkeley-db@18.1.40%gcc@11.2.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-centos7-haswell
[+]              ^gdbm@1.23%gcc@11.2.0 build_system=autotools arch=linux-centos7-haswell
[+]      ^boost@1.77.0%gcc@11.2.0+atomic+chrono~clanglibcpp~container~context~contract~coroutine+date_time~debug+exception~fiber+filesystem+graph~
graph_parallel~icu+iostreams~json+locale+log+math~mpi+multithreaded~nowide~numpy~pic+program_options~python+random+regex+serialization+shared+sign
als~singlethreaded~stacktrace+system~taggedlayout+test+thread+timer~type_erasure~versionedlayout+wave build_system=generic cxxstd=14 patches=48496
71,a440f96 visibility=hidden arch=linux-centos7-haswell
[+]          ^bzip2@1.0.8%gcc@11.2.0~debug~pic+shared build_system=generic arch=linux-centos7-haswell
[+]          ^xz@5.4.1%gcc@11.2.0~pic build_system=autotools libs=shared,static arch=linux-centos7-haswell
[+]          ^zlib-ng@2.1.4%gcc@11.2.0+compat+opt build_system=autotools arch=linux-centos7-haswell
[+]          ^zstd@1.5.5%gcc@11.2.0+programs build_system=makefile compression=none libs=shared,static arch=linux-centos7-haswell
[+]      ^cmake@3.27.7%gcc@11.2.0~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-centos7-haswell
[+]          ^curl@8.4.0%gcc@11.2.0~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=openssl arch=
linux-centos7-haswell
[+]              ^nghttp2@1.57.0%gcc@11.2.0 build_system=autotools arch=linux-centos7-haswell
[+]      ^doxygen@1.9.6%gcc@11.2.0~graphviz~ipo~mscgen build_system=cmake build_type=Release generator=make arch=linux-centos7-haswell
[+]          ^flex@2.6.3%gcc@11.2.0+lex~nls build_system=autotools arch=linux-centos7-haswell
[+]              ^findutils@4.9.0%gcc@11.2.0 build_system=autotools patches=440b954 arch=linux-centos7-haswell
[+]          ^libiconv@1.17%gcc@11.2.0 build_system=autotools libs=shared,static arch=linux-centos7-haswell
[+]          ^python@3.11.6%gcc@11.2.0+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl
~tkinter+uuid+zlib build_system=generic patches=13fa8bf,b0615b2,ebdca64,f2fd060 arch=linux-centos7-haswell
[+]              ^expat@2.5.0%gcc@11.2.0+libbsd build_system=autotools arch=linux-centos7-haswell
[+]                  ^libbsd@0.11.7%gcc@11.2.0 build_system=autotools arch=linux-centos7-haswell
[+]                      ^libmd@1.0.4%gcc@11.2.0 build_system=autotools arch=linux-centos7-haswell
[+]              ^libffi@3.4.4%gcc@11.2.0 build_system=autotools arch=linux-centos7-haswell
[+]              ^libxcrypt@4.4.35%gcc@11.2.0~obsolete_api build_system=autotools patches=4885da3 arch=linux-centos7-haswell
[+]              ^readline@8.2%gcc@11.2.0 build_system=autotools patches=bbf97f1 arch=linux-centos7-haswell
[+]              ^sqlite@3.43.2%gcc@11.2.0+column_metadata+dynamic_extensions+fts~functions+rtree build_system=autotools arch=linux-centos7-haswel
l
[+]              ^util-linux-uuid@2.38.1%gcc@11.2.0 build_system=autotools arch=linux-centos7-haswell
[+]      ^gmake@4.4.1%gcc@11.2.0~guile build_system=generic arch=linux-centos7-haswell
[+]      ^libedit@3.1-20210216%gcc@11.2.0 build_system=autotools arch=linux-centos7-haswell
[+]      ^libtirpc@1.3.3%gcc@11.2.0 build_system=autotools arch=linux-centos7-haswell
[+]          ^krb5@1.20.1%gcc@11.2.0+shared build_system=autotools arch=linux-centos7-haswell
[+]      ^ncurses@6.4%gcc@11.2.0~symlinks+termlib abi=none build_system=autotools arch=linux-centos7-haswell
[+]      ^openssl@3.1.3%gcc@11.2.0~docs+shared build_system=generic certs=mozilla arch=linux-centos7-haswell
[+]          ^ca-certificates-mozilla@2023-05-30%gcc@11.2.0 build_system=generic arch=linux-centos7-haswell
[+]      ^pkgconf@1.9.5%gcc@11.2.0 build_system=autotools arch=linux-centos7-haswell
[+]      ^rpcsvc-proto@1.4.3%gcc@11.2.0 build_system=autotools arch=linux-centos7-haswell
```


### Error message

<details><summary>Error message</summary>

<pre>
...
</pre></details>
==> Error: ProcessError: Command exited with status 1:
...
1 error found in build log:
     377    Make sure you have specified a supported SSL version.
     378    Valid options are :
     379    system (use the OS openssl library),
     380    yes (synonym for system),
     381    </path/to/custom/openssl/installation>
     382    
  >> 383    CMake Error at cmake/ssl.cmake:73 (MESSAGE):
     384      Please install the appropriate openssl developer package.
     385    
     386    Call Stack (most recent call first):
     387      cmake/ssl.cmake:331 (FATAL_SSL_NOT_FOUND_ERROR)
     388      CMakeLists.txt:1688 (MYSQL_CHECK_SSL)

### Information on your system

* **Spack:** 0.21.0.dev0
* **Python:** 3.11.6
* **Platform:** linux-centos7-haswell
* **Concretizer:** clingo

### Additional information


  cmake/boost.cmake:206 (COULD_NOT_FIND_BOOST)
  CMakeLists.txt:1467 (INCLUDE)
-- Configuring incomplete, errors occurred!
(py3.11) -bash-4.2$tail -n35 /tmp/nma8kk/spack-stage/spack-stage-mysql-8.0.29-fv3qxkcg5zwkzet36mfmkdaaqmxadrpz/spack-build-out.txt
-- Performing Test C_LD_LLD_RESULT
-- Performing Test C_LD_LLD_RESULT - Failed
-- Performing Test CXX_LD_LLD_RESULT
-- Performing Test CXX_LD_LLD_RESULT - Failed
-- Performing Test C_LD_GOLD_RESULT
-- Performing Test C_LD_GOLD_RESULT - Success
-- Performing Test CXX_LD_GOLD_RESULT
-- Performing Test CXX_LD_GOLD_RESULT - Success
-- WITH_BOOST=/sfs/qumulo/qhome/nma8kk/libs/spack/opt/spack/linux-centos7-haswell/gcc-11.2.0/boost-1.77.0-gbxkmzaet4guf3ynw3bsunzqiqub5epr
-- BOOST_INCLUDE_DIR 
-- LOCAL_BOOST_DIR LOCAL_BOOST_DIR-NOTFOUND
-- LOCAL_BOOST_ZIP LOCAL_BOOST_ZIP-NOTFOUND
-- Could not find (the correct version of) boost.
-- MySQL currently requires boost_1_77_0
CMake Error at cmake/boost.cmake:108 (MESSAGE):
  You can download it with -DDOWNLOAD_BOOST=1 -DWITH_BOOST=<directory>
  This CMake script will look for boost in <directory>.  If it is not there,
  it will download and unpack it (in that directory) for you.
  You can also download boost manually, from
  https://boostorg.jfrog.io/artifactory/main/release/1.77.0/source/boost_1_77_0.tar.bz2
  If you are inside a firewall, you may need to use an https proxy:
  export https_proxy=http://example.com:80
Call Stack (most recent call first):
  cmake/boost.cmake:206 (COULD_NOT_FIND_BOOST)
  CMakeLists.txt:1467 (INCLUDE)
-- Configuring incomplete, errors occurred!

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

Thank you for pointing out the fix. 
However, once it's happy with boost, it complains about SSL.

>  -- ZSTD_INCLUDE_DIR /cfdbuild/cfdapps/spack-stage/spack-stage-mysql-8.0.29-oxctlg3art3lpnhirqisa2jmrqz3zltx/spack-src/extra/zstd/zstd-1.5.0/lib
-- ZSTD_LIBRARY zstd
-- OPENSSL_INCLUDE_DIR = /mrltools/spack/opt/linux-rocky8-x86_64/gcc-13.2.0/openssl/3.1.3-lzpftsskiru6hzb4jaclv7lqrexwgxd3/include
-- OPENSSL_LIBRARY = /mrltools/spack/opt/linux-rocky8-x86_64/gcc-13.2.0/openssl/3.1.3-lzpftsskiru6hzb4jaclv7lqrexwgxd3/lib64/libssl.so
-- CRYPTO_LIBRARY = /mrltools/spack/opt/linux-rocky8-x86_64/gcc-13.2.0/openssl/3.1.3-lzpftsskiru6hzb4jaclv7lqrexwgxd3/lib64/libcrypto.so
-- OPENSSL_MAJOR_VERSION = 
-- OPENSSL_MINOR_VERSION = 
-- OPENSSL_FIX_VERSION = 
-- Looking for SHA512_DIGEST_LENGTH
-- Looking for SHA512_DIGEST_LENGTH - found
 -- 
Cannot find appropriate system libraries for WITH_SSL=system.
Make sure you have specified a supported SSL version. 
Valid options are : 
system (use the OS openssl library), 
yes (synonym for system), 
</path/to/custom/openssl/installation>

> CMake Error at cmake/ssl.cmake:73 (MESSAGE):
  Please install the appropriate openssl developer package.

> Call Stack (most recent call first):
  cmake/ssl.cmake:331 (FATAL_SSL_NOT_FOUND_ERROR)
  CMakeLists.txt:1688 (MYSQL_CHECK_SSL)


* **Spack:** 0.21.0 (65d3221a9c436e77af9e802e543c71bd702ff4e2)
* **Python:** 3.6.8
* **Platform:** linux-rocky8-x86_64
* **Concretizer:** clingo

This happens with both mysql@8.0.29 and 8.0.19 (I haven't tried any other)

(sorry for the bad formatting)

Now i'm stuck with this.

```console
     1667    [ 51%] Building CXX object libbinlogevents/src/CMakeFiles/binlogevents_static.dir/gtids/gtid.cpp.o
     1668    cd /cfdbuild/cfdapps/spack-stage/spack-stage-mysql-8.0.29-uqvgoijslbf5sp2ks3qgvamcfnbacs6p/spack-build-uqvgoij/libbinlogevents/src && /mrltools/spack/0.21.0/lib/spack/env/gcc/g++ -DHAVE_CONFIG_H -DHAVE_TLSv13 -DLZ4_DISABLE_DEPRECATE_WARNINGS -DRAPIDJSON_NO_SIZETYPEDEF
             INE -DRAPIDJSON_SCHEMA_USE_INTERNALREGEX=0 -DRAPIDJSON_SCHEMA_USE_STDREGEX=1 -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_USE_MATH_DEFINES -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/cfdbuild/cfdapps/spack-stage/spack-stage-mysql-8.0.29-uqvgoijslbf5sp2ks3qgvamcfnbacs6p
             /spack-build-uqvgoij -I/cfdbuild/cfdapps/spack-stage/spack-stage-mysql-8.0.29-uqvgoijslbf5sp2ks3qgvamcfnbacs6p/spack-build-uqvgoij/include -I/cfdbuild/cfdapps/spack-stage/spack-stage-mysql-8.0.29-uqvgoijslbf5sp2ks3qgvamcfnbacs6p/spack-src -I/cfdbuild/cfdapps/spack-sta
             ge/spack-stage-mysql-8.0.29-uqvgoijslbf5sp2ks3qgvamcfnbacs6p/spack-src/include -I/cfdbuild/cfdapps/spack-stage/spack-stage-mysql-8.0.29-uqvgoijslbf5sp2ks3qgvamcfnbacs6p/spack-src/libbinlogevents/include -isystem /cfdbuild/cfdapps/spack-stage/spack-stage-mysql-8.0.29-u
             qvgoijslbf5sp2ks3qgvamcfnbacs6p/spack-src/extra/rapidjson/include -isystem /cfdbuild/cfdapps/spack-stage/spack-stage-mysql-8.0.29-uqvgoijslbf5sp2ks3qgvamcfnbacs6p/spack-src/extra/lz4/lz4-1.9.3 -isystem /cfdbuild/cfdapps/spack-stage/spack-stage-mysql-8.0.29-uqvgoijslbf
             5sp2ks3qgvamcfnbacs6p/spack-src/extra/zstd/zstd-1.5.0/lib -isystem /cfdbuild/cfdapps/spack-stage/spack-stage-mysql-8.0.29-uqvgoijslbf5sp2ks3qgvamcfnbacs6p/spack-build-uqvgoij/extra/zlib -isystem /cfdbuild/cfdapps/spack-stage/spack-stage-mysql-8.0.29-uqvgoijslbf5sp2ks3
             qgvamcfnbacs6p/spack-src/extra/zlib -isystem /mrltools/spack/opt/linux-rocky8-x86_64/gcc-13.2.0/libedit/3.1-20210216-aiymojzmd6s7becniq7v2alzvsqmzjfn/include/editline -std=c++17 -fno-omit-frame-pointer -ftls-model=initial-exec -std=c++17 -Wno-deprecated-declarations -
             Wno-error=register -Wall -Wextra -Wformat-security -Wvla -Wundef -Wmissing-format-attribute -Woverloaded-virtual -Wcast-qual -Wimplicit-fallthrough=5 -Wstringop-truncation -Wsuggest-override -Wmissing-include-dirs -Wextra-semi -Wlogical-op -ffunction-sections -fdata-s
             ections -O3 -DNDEBUG -fPIC -Wshadow=local  -Wno-missing-profile -MD -MT libbinlogevents/src/CMakeFiles/binlogevents_static.dir/gtids/gtid.cpp.o -MF CMakeFiles/binlogevents_static.dir/gtids/gtid.cpp.o.d -o CMakeFiles/binlogevents_static.dir/gtids/gtid.cpp.o -c /cfdbuil
             d/cfdapps/spack-stage/spack-stage-mysql-8.0.29-uqvgoijslbf5sp2ks3qgvamcfnbacs6p/spack-src/libbinlogevents/src/gtids/gtid.cpp
     1669    In file included from /cfdbuild/cfdapps/spack-stage/spack-stage-mysql-8.0.29-uqvgoijslbf5sp2ks3qgvamcfnbacs6p/spack-src/libbinlogevents/include/gtids/gtid.h:29,
     1670                     from /cfdbuild/cfdapps/spack-stage/spack-stage-mysql-8.0.29-uqvgoijslbf5sp2ks3qgvamcfnbacs6p/spack-src/libbinlogevents/src/gtids/gtid.cpp:23:
  >> 1671    /cfdbuild/cfdapps/spack-stage/spack-stage-mysql-8.0.29-uqvgoijslbf5sp2ks3qgvamcfnbacs6p/spack-src/libbinlogevents/include/gtids/global.h:32:14: error: 'int64_t' in namespace 'std' does not name a type
     1672       32 | typedef std::int64_t gno_t;
     1673          |              ^~~~~~~

```

I'm trying to compile with gcc@13.2.0

--- a/var/spack/repos/builtin/packages/mysql/package.py
+++ b/var/spack/repos/builtin/packages/mysql/package.py
@@ -16,6 +16,7 @@ class Mysql(CMakePackage):
     homepage = "https://www.mysql.com/"
     url = "https://dev.mysql.com/get/Downloads/MySQL-8.0/mysql-8.0.15.tar.gz"
 
+    version("8.0.35", sha256="917c5ed38704e99211185ce4be24e33a8c19c91241ed73af4181a6f38d1574c2")
     version("8.0.29", sha256="512170fa6f78a694d6f18d197e999d2716ee68dc541d7644dd922a3663407266")
     version("8.0.19", sha256="a62786d67b5e267eef928003967b4ccfe362d604b80f4523578e0688f5b9f834")
     version("8.0.18", sha256="4cb39a315298eb243c25c53c184b3682b49c2a907a1d8432ba0620534806ade8")
@@ -52,7 +53,7 @@ class Mysql(CMakePackage):
     variant("client_only", default=False, description="Build and install client only.")
     variant(
         "cxxstd",
-        default="14",
+        default="17",
         values=("98", "11", "14", "17"),
         multi=False,
         description="Use the specified C++ standard when building.",
@@ -67,6 +68,7 @@ class Mysql(CMakePackage):
     provides("mysql-client")
 
     # https://dev.mysql.com/doc/refman/8.0/en/source-installation.html
+    # https://dev.mysql.com/doc/refman/8.0/en/source-configuration-options.html
 
     # See CMAKE_MINIMUM_REQUIRED in CMakeLists.txt
     depends_on("cmake@3.1.0:", type="build", when="@5.7.0:5.7 platform=win32")
@@ -83,6 +85,8 @@ class Mysql(CMakePackage):
 
     # Each version of MySQL requires a specific version of boost
     # See BOOST_PACKAGE_NAME in cmake/boost.cmake
+    # 8.0.35
+    depends_on("boost@1.77.0 cxxstd=17", type="build", when="@8.0.35 cxxstd=17")
     # 8.0.29
     depends_on("boost@1.77.0 cxxstd=98", type="build", when="@8.0.29 cxxstd=98")
     depends_on("boost@1.77.0 cxxstd=11", type="build", when="@8.0.29 cxxstd=11")
@@ -124,9 +128,21 @@ class Mysql(CMakePackage):
     # See https://github.com/spack/spack/pull/22303 for reference
     depends_on(Boost.with_default_variants, when="@5.7:")
 
+    with when("@8.0.35:"):
+        depends_on("openssl@3:")
+        requires("cxxstd=17")
+
+    depends_on("zstd", when="@8.0.18:")
+
+    depends_on("patchelf", type="build")
+    depends_on("curl")
+    depends_on("zlib-api")
+    depends_on("libevent")
+    depends_on("lz4")
+
     depends_on("rpcsvc-proto")
     depends_on("ncurses")
-    depends_on("openssl")
+    depends_on("openssl@:2", when="@:8.0.29")
     depends_on("libtirpc", when="@5.7.0: platform=linux")
     depends_on("libedit", type=["build", "run"])
     depends_on("perl", type=["build", "test"], when="@:7")
@@ -150,14 +166,28 @@ class Mysql(CMakePackage):
 
     def cmake_args(self):
         spec = self.spec
-        options = []
-        if "boost" in spec:
-            options.append("-DBOOST_ROOT={0}".format(spec["boost"].prefix))
-        if "+client_only" in self.spec:
-            options.append("-DWITHOUT_SERVER:BOOL=ON")
-        options.append("-DWITH_EDITLINE=system")
-        options.append("-Dlibedit_INCLUDE_DIR={0}".format(spec["libedit"].prefix.include))
-        options.append("-Dlibedit_LIBRARY={0}".format(spec["libedit"].libs.directories[0]))
+        options = [
+            self.define("REPRODUCIBLE_BUILD", True),
+            self.define("WITH_CURL", spec["curl"].prefix),
+            self.define("WITH_EDITLINE", "system"),
+            self.define("WITH_LIBEVENT", "system"),
+            self.define("WITH_LZ4", "system"),
+            self.define("WITH_SSL", spec["openssl"].prefix),
+            self.define("WITH_ZLIB", "system"),
+            self.define_from_variant("WITHOUT_SERVER", "client_only"),
+        ]
+
+        if spec.satisfies("@5.7:"):
+            options.extend(
+                [
+                    self.define("WITH_BOOST", spec["boost"].prefix),
+                    self.define("LOCAL_BOOST_DIR", spec["boost"].prefix),
+                ]
+            )
+
+        if spec.satisfies("@8.0.18:"):
+            options.extend([self.define("WITH_ZSTD", "system")])
+
         return options
 
     def _fix_dtrace_shebang(self, env):

